src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/court/VideoLinkMigrationService.kt

```diff
@@ -0,0 +1,144 @@
+      require((appointments.size <= 3)) { "Video link booking ID $videoBookingId has more than 3 appointments" }
+    }
+
+    // Extract the related appointments - will be null if a cancelled booking otherwise reflect the appointments
+    val main = appointments.find { it.hearingType == HearingType.MAIN }
+    val pre = appointments.find { it.hearingType == HearingType.PRE }
+    val post = appointments.find { it.hearingType == HearingType.POST }
+
+    // Extract appointment details from the latest CREATE or UPDATE event if this booking is cancelled
+    val appointmentsFromEvent = extractAppointmentsFromEvent(eventEntities, cancelledBooking)
+
+    if (!cancelledBooking) {
+      require(main != null) { "Video link booking ID $videoBookingId has no main appointment" }
+    }
+
+    return VideoBookingMigrateResponse(
+      videoBookingId = videoLinkBooking.id!!,
+      offenderBookingId = videoLinkBooking.offenderBookingId,
+      courtCode = videoLinkBooking.courtId?.let { videoLinkBooking.courtId } ?: "UNKNOWN",
+      courtName = videoLinkBooking.courtName,
+      madeByTheCourt = videoLinkBooking.madeByTheCourt?.let { videoLinkBooking.madeByTheCourt } ?: false,
+      createdByUsername = videoLinkBooking.createdByUsername?.let { videoLinkBooking.createdByUsername } ?: "MIGRATED",
+      prisonCode = videoLinkBooking.prisonId,
+      probation = videoLinkBooking.courtName?.contains("Probation") ?: false,
+      cancelled = cancelledBooking,
+      comment = videoLinkBooking.comment,
+      pre = if (!cancelledBooking) pre?.let { mapAppointment(pre) } else appointmentsFromEvent.pre,
+      main = if (!cancelledBooking) mapAppointment(main!!) else appointmentsFromEvent.main!!,
+      post = if (!cancelledBooking) post?.let { mapAppointment(post) } else appointmentsFromEvent.post,
+      events = mapEvents(eventEntities),
+    )
+  }
+
+  fun mapAppointment(appointment: VideoLinkAppointment) =
+    AppointmentLocationTimeSlot(
+      locationId = appointment.locationId,
+      date = LocalDate.from(appointment.startDateTime),
+      startTime = LocalTime.from(appointment.startDateTime).withSecond(0).withNano(0),
+      endTime = LocalTime.from(appointment.endDateTime).withSecond(0).withNano(0),
+    )
+
+  fun mapEventToLocationTimeSlot(locationId: Long, startTime: LocalDateTime, endTime: LocalDateTime) =
+    AppointmentLocationTimeSlot(
+      locationId = locationId,
+      date = LocalDate.of(startTime.year, startTime.month, startTime.dayOfMonth),
+      startTime = LocalTime.of(startTime.hour, startTime.minute),
+      endTime = LocalTime.of(endTime.hour, endTime.minute),
+    )
+
+  private fun mapEvents(events: List<VideoLinkBookingEvent>): List<VideoBookingMigrateEvent> {
+    val response = events.map { event ->
+      VideoBookingMigrateEvent(
+        eventId = event.eventId!!,
+        eventTime = event.timestamp,
+        eventType = event.eventType,
+        createdByUsername = event.userId?.let { event.userId } ?: "MIGRATED",
```

Comment:
Similar here with the use of `.let { ... }`'s, not needed.

Reply:
Done